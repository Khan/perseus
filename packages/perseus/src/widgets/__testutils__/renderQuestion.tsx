import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
import {render} from "@testing-library/react";
import * as React from "react";

import {
    testDependenciesV2,
    testDependencies,
    // eslint-disable-next-line import/no-relative-packages
} from "../../../../../testing/test-dependencies";
import {
    DependenciesContext,
    useDependencies,
    setDependencies,
} from "../../dependencies";
import * as Perseus from "../../index";
import {mockStrings} from "../../strings";
import {registerAllWidgetsForTesting} from "../../util/register-all-widgets-for-testing";

import type {APIOptions} from "../../types";
import type {PerseusRenderer} from "@khanacademy/perseus-core";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type RenderResult = ReturnType<typeof render>;

type ExtraProps = Omit<PropsFor<typeof Perseus.Renderer>, "strings">;

export const renderQuestion = (
    question: PerseusRenderer,
    apiOptions: APIOptions = Object.freeze({}),
    extraProps?: ExtraProps,
): {
    container: HTMLElement;
    renderer: Perseus.Renderer;
    rerender: (question: PerseusRenderer, extraProps?: ExtraProps) => void;
    unmount: RenderResult["unmount"];
} => {
    setDependencies(testDependencies);
    registerAllWidgetsForTesting();

    let renderer: Perseus.Renderer | null = null;
    const {container, rerender, unmount} = render(
        <RenderStateRoot>
            <DependenciesContext.Provider value={testDependenciesV2}>
                <Renderer
                    ref={(node) => (renderer = node)}
                    question={question as any}
                    apiOptions={apiOptions}
                    extraProps={{
                        ...extraProps,
                        strings: mockStrings,
                    }}
                />
            </DependenciesContext.Provider>
        </RenderStateRoot>,
    );
    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
    if (!renderer) {
        throw new Error(`Failed to render!`);
    }
    const renderAgain = (
        question: PerseusRenderer,
        extraProps?: ExtraProps,
    ) => {
        rerender(
            <RenderStateRoot>
                <DependenciesContext.Provider value={testDependenciesV2}>
                    <Renderer
                        ref={(node) => (renderer = node)}
                        question={question}
                        apiOptions={apiOptions}
                        extraProps={{
                            ...extraProps,
                            strings: mockStrings,
                        }}
                    />
                </DependenciesContext.Provider>
            </RenderStateRoot>,
        );
        if (!renderer) {
            throw new Error(`Failed to rerender!`);
        }
    };

    return {container, renderer, rerender: renderAgain, unmount};
};

const Renderer = React.forwardRef<
    Perseus.Renderer,
    {
        question: PerseusRenderer;
        apiOptions: APIOptions;
        extraProps?: PropsFor<typeof Perseus.Renderer>;
    }
>((props, ref) => {
    const dependencies = useDependencies();
    return (
        <Perseus.Renderer
            ref={ref}
            content={props.question.content}
            images={props.question.images}
            widgets={props.question.widgets}
            problemNum={0}
            apiOptions={props.apiOptions}
            strings={mockStrings}
            {...props.extraProps}
            {...dependencies}
        />
    );
});
