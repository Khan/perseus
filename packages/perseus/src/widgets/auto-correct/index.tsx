import React from "react";

import type {PerseusScore, WidgetExports} from "../../types";

// The props are type `any` on purpose so that this can receive props
// from any deprecated widget
type AutoCorrectProps = any;
type Rubric = any;
type UserInput = Empty;

class AutoCorrect extends React.Component<AutoCorrectProps> {
    static validate(userInput: UserInput, rubric: Rubric): PerseusScore {
        return {
            type: "points",
            earned: 0,
            total: 0,
            message: null,
        };
    }

    getUserInput: () => UserInput = () => {
        return {};
    };

    simpleValidate: (arg1: Rubric) => PerseusScore = (rubric) => {
        return AutoCorrect.validate(this.getUserInput(), rubric);
    };

    render() {
        return (
            <div
                style={{
                    backgroundColor: "gray",
                    border: "1px solid red",
                }}
            >
                <h1>Gone</h1>
                <p>The widget you are trying to reach is no longer available</p>
            </div>
        );
    }
}

export default {
    name: "auto-correct",
    displayName: "Auto Program",
    widget: AutoCorrect,
} as WidgetExports<typeof AutoCorrect>;
