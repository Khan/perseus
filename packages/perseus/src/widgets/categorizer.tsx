/* eslint-disable @khanacademy/ts-no-error-suppressions */
/* eslint-disable react/sort-comp */
import {linterContextDefault} from "@khanacademy/perseus-linter";
import * as i18n from "@khanacademy/wonder-blocks-i18n";
import {StyleSheet, css} from "aphrodite";
import classNames from "classnames";
import * as React from "react";
import _ from "underscore";

import InlineIcon from "../components/inline-icon";
import {iconCircle, iconCircleThin} from "../icon-paths";
import * as Changeable from "../mixins/changeable";
import WidgetJsonifyDeprecated from "../mixins/widget-jsonify-deprecated";
import {ClassNames as ApiClassNames} from "../perseus-api";
import Renderer from "../renderer";
import mediaQueries from "../styles/media-queries";
import sharedStyles from "../styles/shared";
import Util from "../util";

import type {PerseusCategorizerWidgetOptions} from "../perseus-types";
import type {PerseusScore, WidgetExports, WidgetProps} from "../types";

type UserInput = any;

export type Rubric = PerseusCategorizerWidgetOptions;

type Props = WidgetProps<RenderProps, Rubric> & {
    values: ReadonlyArray<string>;
};

type DefaultProps = {
    items: Props["items"];
    categories: Props["categories"];
    values: Props["values"];
    linterContext: Props["linterContext"];
};

type State = {
    uniqueId: string;
};

export class Categorizer extends React.Component<Props, State> {
    static defaultProps: DefaultProps = {
        items: [],
        categories: [],
        values: [],
        linterContext: linterContextDefault,
    };

    state: State = {
        uniqueId: _.uniqueId("perseus_radio_"),
    };

    change: (...args: ReadonlyArray<unknown>) => any = (...args) => {
        // @ts-expect-error - TS2345 - Argument of type 'readonly unknown[]' is not assignable to parameter of type 'any[]'.
        return Changeable.change.apply(this, args);
    };

    getUserInput: () => UserInput = () => {
        return Categorizer.getUserInputFromProps(this.props);
    };

    render(): React.ReactNode {
        const self = this;

        // In this context, isMobile is used to differentiate mobile from
        // desktop.
        const isMobile = this.props.apiOptions.isMobile;
        let indexedItems = this.props.items.map((item, n) => [item, n]);
        if (this.props.randomizeItems) {
            // @ts-expect-error - TS4104 - The type 'readonly (string | number)[][]' is 'readonly' and cannot be assigned to the mutable type '(string | number)[][]'. | TS2345 - Argument of type 'number | null | undefined' is not assignable to parameter of type 'number | RNG'.
            indexedItems = Util.shuffle(indexedItems, this.props.problemNum);
        }

        const table = (
            <table className={"categorizer-table " + css(styles.mobileTable)}>
                <thead>
                    <tr>
                        <td className={css(styles.emptyHeaderCell)} />
                        {this.props.categories.map((category, i) => {
                            // Array index is the correct key here, as that's
                            // how category grading actually works -- no way
                            // to add or remove categories or items in the
                            // middle. (If we later add that, this should be
                            // fixed.)
                            return (
                                <th className={css(styles.header)} key={i}>
                                    <Renderer
                                        content={category}
                                        linterContext={this.props.linterContext}
                                    />
                                </th>
                            );
                        })}
                    </tr>
                </thead>
                <tbody>
                    {indexedItems.map((indexedItem) => {
                        const item = indexedItem[0];
                        const itemNum = indexedItem[1];
                        const uniqueId = self.state.uniqueId + "_" + itemNum;
                        return (
                            <tr key={itemNum}>
                                <td>
                                    <Renderer
                                        // @ts-expect-error - TS2322 - Type 'string | number' is not assignable to type 'string | undefined'.
                                        content={item}
                                        linterContext={this.props.linterContext}
                                    />
                                </td>
                                {self.props.categories.map(
                                    (catName, catNum) => {
                                        const selected =
                                            self.props.values[itemNum] ===
                                            catNum;
                                        return (
                                            <td
                                                className={
                                                    "category " +
                                                    css(styles.cell)
                                                }
                                                key={catNum}
                                            >
                                                {/* a pseudo-label: toggle the
                                value of the checkbox when this div or the
                                checkbox is clicked */}
                                                <div
                                                    className={
                                                        ApiClassNames.INTERACTIVE
                                                    }
                                                    role="button"
                                                    aria-label={catName}
                                                    // eslint-disable-next-line react/jsx-no-bind
                                                    onClick={this.onChange.bind(
                                                        this,
                                                        itemNum,
                                                        catNum,
                                                    )}
                                                >
                                                    {isMobile && (
                                                        <input
                                                            type="radio"
                                                            name={uniqueId}
                                                            className={css(
                                                                sharedStyles.responsiveInput,
                                                                sharedStyles.responsiveRadioInput,
                                                            )}
                                                            checked={selected}
                                                            // eslint-disable-next-line react/jsx-no-bind
                                                            onChange={this.onChange.bind(
                                                                this,
                                                                itemNum,
                                                                catNum,
                                                            )}
                                                            onClick={(e) =>
                                                                e.stopPropagation()
                                                            }
                                                        />
                                                    )}
                                                    {!isMobile && (
                                                        <span
                                                            className={css(
                                                                styles.radioSpan,
                                                                selected &&
                                                                    styles.checkedRadioSpan,
                                                                this.props
                                                                    .static &&
                                                                    selected &&
                                                                    styles.staticCheckedRadioSpan,
                                                            )}
                                                        >
                                                            {selected ? (
                                                                <InlineIcon
                                                                    {...iconCircle}
                                                                />
                                                            ) : (
                                                                <InlineIcon
                                                                    {...iconCircleThin}
                                                                />
                                                            )}
                                                        </span>
                                                    )}
                                                </div>
                                            </td>
                                        );
                                    },
                                )}
                            </tr>
                        );
                        /* eslint-enable max-len */
                    })}
                </tbody>
            </table>
        );

        // TODO(benkomalo): kill CSS-based styling and move everything to
        // aphrodite.
        const extraClassNames = classNames({
            "categorizer-container": true,
            "static-mode": this.props.static,
        });
        const inlineStyles = this.props.apiOptions.isMobile
            ? [styles.fullBleedContainer]
            : [];

        return (
            <div className={extraClassNames + " " + css(...inlineStyles)}>
                {table}
            </div>
        );
    }

    onChange: (arg1: number, arg2: number) => void = (itemNum, catNum) => {
        const values = [...this.props.values];
        // @ts-expect-error - TS2322 - Type 'number' is not assignable to type 'never'.
        values[itemNum] = catNum;
        this.change("values", values);
        this.props.trackInteraction();
    };

    simpleValidate: (arg1: Rubric) => PerseusScore = (rubric) => {
        return Categorizer.validate(this.getUserInput(), rubric);
    };

    static validate(userInput: UserInput, rubric: Rubric): PerseusScore {
        let completed = true;
        let allCorrect = true;
        _.each(rubric.values, function (value, i) {
            if (userInput.values[i] == null) {
                completed = false;
            }
            if (userInput.values[i] !== value) {
                allCorrect = false;
            }
        });
        if (!completed) {
            return {
                type: "invalid",
                message: i18n._(
                    "Make sure you select something for every row.",
                ),
            };
        }
        return {
            type: "points",
            earned: allCorrect ? 1 : 0,
            total: 1,
            message: null,
        };
    }

    static getUserInputFromProps(props: Props): UserInput {
        return WidgetJsonifyDeprecated.getUserInputFromProps(props);
    }
}

// TODO(benkomalo): inject page-margin into Perseus instead of hardcoding.
const pageMargin = 16;

// Stylesheets aren't directly testable
/* c8 ignore next */
const styles = StyleSheet.create({
    mobileTable: {
        [mediaQueries.smOrSmaller]: {
            minWidth: "auto",
        },
    },

    fullBleedContainer: {
        [mediaQueries.mdOrSmaller]: {
            marginLeft: -pageMargin,
            marginRight: -pageMargin,
            overflowX: "auto",
        },
    },

    header: {
        textAlign: "center",
        verticalAlign: "bottom",
    },

    cell: {
        textAlign: "center",
        padding: 0,
        color: "#ccc",
        verticalAlign: "middle",
    },

    emptyHeaderCell: {
        backgroundColor: "inherit",
        borderBottom: "2px solid #ccc",
    },

    radioSpan: {
        fontSize: 30,
        paddingRight: 3,

        ":hover": {
            color: "#999",
        },
    },

    checkedRadioSpan: {
        color: "#333",
    },

    // .static-mode is applied by the Categorizer when the rendered
    // widget is static; in this case we gray out the choices to show
    // the user that the widget can't be interacted with.
    staticCheckedRadioSpan: {
        color: "#888",
    },
});

type RenderProps = {
    items: PerseusCategorizerWidgetOptions["items"];
    categories: PerseusCategorizerWidgetOptions["categories"];
    randomizeItems: PerseusCategorizerWidgetOptions["randomizeItems"];
    // Depends on whether the widget is in static mode
    values?: PerseusCategorizerWidgetOptions["values"];
};

export default {
    name: "categorizer",
    displayName: "Categorizer",
    widget: Categorizer,
    transform: (
        widgetOptions: PerseusCategorizerWidgetOptions,
    ): RenderProps => {
        return _.pick(widgetOptions, "items", "categories", "randomizeItems");
    },
    staticTransform: (
        editorProps: PerseusCategorizerWidgetOptions,
    ): RenderProps => {
        return _.pick(
            editorProps,
            "items",
            "categories",
            "values",
            "randomizeItems",
        );
    },
    isLintable: true,
} as WidgetExports<typeof Categorizer>;
