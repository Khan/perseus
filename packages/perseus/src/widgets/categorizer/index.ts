export {default, Categorizer} from "./categorizer";
