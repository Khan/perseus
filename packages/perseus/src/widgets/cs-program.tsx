/**
 * This widget is for embedding Khan Academy CS programs.
 */

import {StyleSheet, css} from "aphrodite";
import $ from "jquery";
import PropTypes from "prop-types";
import * as React from "react";
import _ from "underscore";

import {getDependencies} from "../dependencies";
import * as Changeable from "../mixins/changeable";
import {articleMaxWidthInPx} from "../styles/constants";
import Util from "../util";
import {isFileProtocol} from "../util/mobile-native-utils";
import {toAbsoluteUrl} from "../util/url-utils";

import type {WidgetExports} from "../types";

const {updateQueryString} = Util;

function getUrlFromProgramID(programID: any) {
    const {InitialRequestUrl} = getDependencies();

    const path =
        "/computer-programming/program/" +
        `${programID}/embedded?embed=yes&author=no`;
    // When loading scratchpads in our native apps, the current URL will exist
    // as a `file://` URL. In this case, we want to reference `ka.org` instead
    // of defaulting to what we get from `toAbsoluteUrl`.
    if (isFileProtocol(InitialRequestUrl.protocol)) {
        return `https://khanacademy.org${path}`;
    }

    return toAbsoluteUrl(path);
}

/* This renders the scratchpad in an iframe and handles validation via
 * window.postMessage */
class CSProgram extends React.Component<any> {
    static propTypes = {
        ...Changeable.propTypes,
        programID: PropTypes.string,
        programType: PropTypes.oneOf(["pjs", "sql", "webpage"]),
        width: PropTypes.number,
        height: PropTypes.number,
        // eslint-disable-next-line react/forbid-prop-types
        settings: PropTypes.array,
        showEditor: PropTypes.bool,
        showButtons: PropTypes.bool,
        status: PropTypes.oneOf(["incomplete", "incorrect", "correct"]),
        message: PropTypes.string,
    };

    // The widget's grading function
    static validate(state: any, rubric: any): any {
        // The iframe can tell us whether it's correct or incorrect,
        //  and pass an optional message
        if (state.status === "correct") {
            return {
                type: "points",
                earned: 1,
                total: 1,
                message: state.message || null,
            };
        }
        if (state.status === "incorrect") {
            return {
                type: "points",
                earned: 0,
                total: 1,
                message: state.message || null,
            };
        }
        return {
            type: "invalid",
            message: "Keep going, you're not there yet!",
        };
    }

    static defaultProps: any = {
        showEditor: false,
        showButtons: false,
        status: "incomplete",
        // optional message
        message: null,
    };

    componentDidMount() {
        $(window).on("message", this.handleMessageEvent);
    }

    componentWillUnmount() {
        $(window).off("message", this.handleMessageEvent);
    }

    handleMessageEvent: (arg1: any) => void = (e) => {
        // We receive data from the iframe that contains {passed: true/false}
        //  and use that to set the status
        // It could also contain an optional message
        let data: Record<string, any> = {};
        try {
            data = JSON.parse(e.originalEvent.data);
        } catch (err: any) {
            return;
        }

        if (_.isUndefined(data.testsPassed)) {
            return;
        }

        const status = data.testsPassed ? "correct" : "incorrect";
        this.change({
            status: status,
            message: data.message,
        });
    };

    change: (...args: ReadonlyArray<unknown>) => any = (...args) => {
        // @ts-expect-error - TS2345 - Argument of type 'readonly unknown[]' is not assignable to parameter of type 'any[]'.
        return Changeable.change.apply(this, args);
    };

    simpleValidate: (arg1: any) => any = (rubric) => {
        return CSProgram.validate(
            {
                status: this.props.status,
                message: this.props.message,
            },
            rubric,
        );
    };

    render(): React.ReactNode {
        if (!this.props.programID) {
            return <div />;
        }

        let styleContainer = false;
        let url = getUrlFromProgramID(this.props.programID);
        let className;
        const style = {
            height: this.props.height,
            width: "100%",
        } as const;

        if (this.props.showEditor) {
            url += "&editor=yes";
            className = "perseus-scratchpad-editor";
        } else {
            url += `&editor=no&width=${articleMaxWidthInPx}`;
            className = "perseus-scratchpad";
            if (this.props.programType !== "webpage") {
                styleContainer = true;
            }
        }

        if (this.props.showButtons) {
            url += "&buttons=yes";
            // Matches templates/scratchpads/embed_script.js
            // Toolbar height is 66, border height is 1 pixel
            // @ts-expect-error - TS2540 - Cannot assign to 'height' because it is a read-only property.
            style.height += 67;
        } else {
            url += "&buttons=no";
        }

        // Turn array of [{name: "", value: ""}] into object
        if (this.props.settings) {
            const settings: Record<string, any> = {};
            _.each(this.props.settings, function (setting) {
                if (setting.name && setting.value) {
                    settings[setting.name] = setting.value;
                }
            });
            // This becomes available to programs as Program.settings()
            url = updateQueryString(url, "settings", JSON.stringify(settings));
        }

        const sandboxOptions = [
            "allow-popups",
            "allow-same-origin",
            "allow-scripts",
            "allow-top-navigation",
        ].join(" ");

        // We sandbox the iframe so that we allowlist only the functionality
        //  that we need. This makes it a bit safer in case some content
        //  creator "went wild".
        // http://www.html5rocks.com/en/tutorials/security/sandboxed-iframes/
        return (
            <div
                className={css(
                    styles.widthOverride,
                    styleContainer && styles.container,
                )}
            >
                <iframe
                    sandbox={sandboxOptions}
                    src={url}
                    style={style}
                    className={className}
                    allowFullScreen={true}
                />
            </div>
        );
    }
}

const styles = StyleSheet.create({
    // Override the inherited width from the perseus paragraph class
    // 820 is minimum width for 420px editor and 400px canvas
    widthOverride: {
        width: 820,
    },
    container: {
        margin: "auto",
    },
});

export default {
    name: "cs-program",
    displayName: "CS Program",
    supportedAlignments: ["block", "full-width"],
    widget: CSProgram,
    hidden: true,
} as WidgetExports<typeof CSProgram>;
