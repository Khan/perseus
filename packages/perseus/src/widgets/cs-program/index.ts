export {default} from "./cs-program";
