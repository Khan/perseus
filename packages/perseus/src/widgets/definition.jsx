// @flow

import Button from "@khanacademy/wonder-blocks-button";
import Color from "@khanacademy/wonder-blocks-color";
import {Popover, PopoverContentCore} from "@khanacademy/wonder-blocks-popover";
import * as React from "react";

import {DefinitionConsumer} from "../definition-context.js";
import Renderer from "../renderer.jsx";

import type {
    PerseusRenderer,
    PerseusDefinitionWidgetOptions,
} from "../perseus-types.js";
import type {PerseusScore, WidgetExports, WidgetProps} from "../types.js";

type RenderProps = PerseusDefinitionWidgetOptions;

type Rubric = PerseusDefinitionWidgetOptions;

type UserInput = Empty;

type DefinitionProps = {|
    ...WidgetProps<RenderProps, Rubric>,
    widgets: PerseusRenderer["widgets"],
|};

class Definition extends React.Component<DefinitionProps> {
    static defaultProps: $FlowFixMe = {
        togglePrompt: "define me",
        definition: "definition goes here",
    };

    static validate(userInput: UserInput, rubric: Rubric): PerseusScore {
        return {
            type: "points",
            earned: 0,
            total: 0,
            message: null,
        };
    }

    getUserInput: () => UserInput = () => {
        return {};
    };

    simpleValidate: (Rubric) => PerseusScore = (rubric) => {
        return Definition.validate(this.getUserInput(), rubric);
    };

    render(): React.Node {
        return (
            <DefinitionConsumer>
                {({activeDefinitionId, setActiveDefinitionId}) => (
                    <Popover
                        content={
                            <PopoverContentCore
                                color="white"
                                style={styles.tooltipBody}
                                closeButtonVisible={true}
                            >
                                <Renderer
                                    apiOptions={this.props.apiOptions}
                                    content={this.props.definition}
                                    widgets={this.props.widgets}
                                />
                            </PopoverContentCore>
                        }
                        opened={activeDefinitionId === this.props.widgetId}
                        onClose={() => setActiveDefinitionId(null)}
                        placement="top"
                    >
                        <span className="perseus-widget-definition">
                            <Button
                                size="small"
                                kind="tertiary"
                                onClick={() => {
                                    this.props.trackInteraction();
                                    setActiveDefinitionId(this.props.widgetId);
                                }}
                            >
                                {this.props.togglePrompt}
                            </Button>
                        </span>
                    </Popover>
                )}
            </DefinitionConsumer>
        );
    }
}

const styles = {
    tooltipBody: {
        color: Color.offBlack,
        fontSize: 20,
        fontWeight: 500,
        lineHeight: "30px",
    },
};

export default ({
    name: "definition",
    displayName: "Definition",
    accessible: true,
    defaultAlignment: "inline",
    widget: Definition,
    transform: (x: any) => x,
}: WidgetExports<typeof Definition>);
