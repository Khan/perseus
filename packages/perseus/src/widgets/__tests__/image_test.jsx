// @flow

import * as Dependencies from "../../../perseus-all-package/dependencies.js";
import {isAccessible} from "../../../perseus-all-package/widgets.js";
import {renderQuestion} from "../../../perseus-all-package/widgets/__tests__/renderQuestion.jsx";
import {testDependencies} from "../../../perseus-testing/test-dependencies.js";
import {question} from "../__testdata__/image_testdata.js";

import type {
    ImageWidget,
    PerseusImageWidgetOptions,
} from "../../../perseus-all-package/perseus-types.js";
import type {APIOptions} from "../../../perseus-all-package/types.js";

describe.each([true, false])("image widget - isMobile %b", (isMobile) => {
    const apiOptions: APIOptions = {isMobile};

    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("should snapshot", () => {
        // Arrange

        // Act
        const {container} = renderQuestion(question, apiOptions);

        // Assert
        expect(container).toMatchSnapshot("first render");
    });

    it("should be unanswerable", () => {
        // Arrange

        // Act
        const {renderer} = renderQuestion(question, apiOptions);

        // Assert
        expect(renderer).toHaveBeenAnsweredIncorrectly();
    });

    it("should be accessible if background has 'alt' prop", () => {
        // Arrange, Act, and Assert
        expect(isAccessible(question.widgets["image 1"])).toBeTrue();
    });

    it("should be inaccessible if background is missing 'alt' prop", () => {
        // Arrange
        const imageWidget = ((question.widgets["image 1"]: any): ImageWidget);
        const options = ((imageWidget.options: any): PerseusImageWidgetOptions);
        const inaccessibleWidgetInfo = ({
            ...imageWidget,
            options: {
                ...options,
                alt: "",
            },
        }: ImageWidget);

        // Act and Assert
        expect(isAccessible(inaccessibleWidgetInfo)).toBeFalse();
    });
});
