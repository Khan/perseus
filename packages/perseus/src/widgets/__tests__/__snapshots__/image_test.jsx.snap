// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`image widget - isMobile %b should snapshot: first render 1`] = `
<div>
  <span
    class="mock-KatexProvider"
  >
    <div
      class="perseus-renderer perseus-renderer-responsive perseus-renderer-two-columns"
    >
      <div
        class="paragraph"
        data-perseus-paragraph-index="0"
      >
        <div
          class="perseus-two-columns"
        >
          <div
            class="perseus-column"
          >
            <div
              class="perseus-column-content"
            >
              <div
                class="paragraph"
              >
                <div
                  class="perseus-widget-container widget-nohighlight widget-block"
                >
                  <div
                    class="perseus-image-widget"
                  >
                    <div
                      class="fixed-to-responsive svg-image"
                      style="max-width: 420px; max-height: 345px;"
                    >
                      <div
                        style="padding-bottom: 82.14%;"
                      />
                      <span
                        style="background: url(/images/spinner.gif) no-repeat; background-position: center; width: 100%; height: 100%; position: absolute; min-width: 20px;"
                      />
                    </div>
                    <span
                      class="perseus-sr-only"
                    >
                      <span
                        class="mock-KatexProvider"
                      >
                        <div
                          class="perseus-renderer perseus-renderer-responsive"
                        >
                          <div
                            class="paragraph"
                            data-perseus-paragraph-index="0"
                          >
                            <div
                              class="paragraph"
                            >
                              An array of isosceles triangles. A triangle has height A. Two smaller triangle, one with height B and one with height C, have approximately the same combined height as A.
                            </div>
                          </div>
                        </div>
                      </span>
                    </span>
                    <div
                      class="perseus-image-caption has-title"
                    >
                      <div
                        class="caption_4bstk7"
                      >
                        <span
                          class="mock-KatexProvider"
                        >
                          <div
                            class="perseus-renderer perseus-renderer-responsive"
                          >
                            <div
                              class="paragraph"
                              data-perseus-paragraph-index="0"
                            >
                              <div
                                class="paragraph"
                              >
                                <strong>
                                  Image Title.
                                </strong>
                                 Image Caption
                              </div>
                            </div>
                          </div>
                        </span>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="perseus-column"
          >
            <div
              class="sat-header-grafting-area"
            />
            <div
              class="perseus-column-content"
            >
              <div
                class="sat-skill-subscore-grafting-area"
              />
              <div
                class="paragraph"
              >
                A quilter wants to make the design shown at left using the Golden Ratio. Specifically, he wants the ratio of the triangle heights 
                <span
                  style="white-space: nowrap;"
                >
                  <span />
                  <span
                    class="mock-TeX"
                  >
                    A:B
                  </span>
                  <span />
                </span>
                 and 
                <span
                  style="white-space: nowrap;"
                >
                  <span />
                  <span
                    class="mock-TeX"
                  >
                    B:C
                  </span>
                  <span />
                </span>
                 to each equal 
                <span
                  style="white-space: nowrap;"
                >
                  <span />
                  <span
                    class="mock-TeX"
                  >
                    1.62
                  </span>
                  <span />
                </span>
                . If the quilter makes the triangle height 
                <span
                  style="white-space: nowrap;"
                >
                  <span />
                  <span
                    class="mock-TeX"
                  >
                    A=8\\ \\text{in}
                  </span>
                  <span />
                </span>
                , approximately how tall should he make triangle height 
                <span
                  style="white-space: nowrap;"
                >
                  <span />
                  <span
                    class="mock-TeX"
                  >
                    C
                  </span>
                  <span />
                </span>
                ?
              </div>
              <div
                class="sat-grafting-area"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </span>
</div>
`;

exports[`image widget - isMobile %b should snapshot: first render 2`] = `
<div>
  <span
    class="mock-KatexProvider"
  >
    <div
      class="perseus-renderer perseus-renderer-responsive perseus-renderer-two-columns"
    >
      <div
        class="paragraph"
        data-perseus-paragraph-index="0"
      >
        <div
          class="perseus-two-columns"
        >
          <div
            class="perseus-column"
          >
            <div
              class="perseus-column-content"
            >
              <div
                class="paragraph"
              >
                <div
                  class="perseus-widget-container widget-nohighlight widget-block"
                >
                  <div
                    class="perseus-image-widget"
                  >
                    <div
                      class="perseus-image-title"
                    >
                      <span
                        class="mock-KatexProvider"
                      >
                        <div
                          class="perseus-renderer perseus-renderer-responsive"
                        >
                          <div
                            class="paragraph"
                            data-perseus-paragraph-index="0"
                          >
                            <div
                              class="paragraph"
                            >
                              Image Title
                            </div>
                          </div>
                        </div>
                      </span>
                    </div>
                    <div
                      class="fixed-to-responsive svg-image"
                      style="max-width: 420px; max-height: 345px;"
                    >
                      <div
                        style="padding-bottom: 82.14%;"
                      />
                      <span
                        style="background: url(/images/spinner.gif) no-repeat; background-position: center; width: 100%; height: 100%; position: absolute; min-width: 20px;"
                      />
                    </div>
                    <span
                      class="perseus-sr-only"
                    >
                      <span
                        class="mock-KatexProvider"
                      >
                        <div
                          class="perseus-renderer perseus-renderer-responsive"
                        >
                          <div
                            class="paragraph"
                            data-perseus-paragraph-index="0"
                          >
                            <div
                              class="paragraph"
                            >
                              An array of isosceles triangles. A triangle has height A. Two smaller triangle, one with height B and one with height C, have approximately the same combined height as A.
                            </div>
                          </div>
                        </div>
                      </span>
                    </span>
                    <div
                      class="perseus-image-caption"
                    >
                      <span
                        class="mock-KatexProvider"
                      >
                        <div
                          class="perseus-renderer perseus-renderer-responsive"
                        >
                          <div
                            class="paragraph"
                            data-perseus-paragraph-index="0"
                          >
                            <div
                              class="paragraph"
                            >
                              Image Caption
                            </div>
                          </div>
                        </div>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="perseus-column"
          >
            <div
              class="sat-header-grafting-area"
            />
            <div
              class="perseus-column-content"
            >
              <div
                class="sat-skill-subscore-grafting-area"
              />
              <div
                class="paragraph"
              >
                A quilter wants to make the design shown at left using the Golden Ratio. Specifically, he wants the ratio of the triangle heights 
                <span
                  style="white-space: nowrap;"
                >
                  <span />
                  <span
                    class="mock-TeX"
                  >
                    A:B
                  </span>
                  <span />
                </span>
                 and 
                <span
                  style="white-space: nowrap;"
                >
                  <span />
                  <span
                    class="mock-TeX"
                  >
                    B:C
                  </span>
                  <span />
                </span>
                 to each equal 
                <span
                  style="white-space: nowrap;"
                >
                  <span />
                  <span
                    class="mock-TeX"
                  >
                    1.62
                  </span>
                  <span />
                </span>
                . If the quilter makes the triangle height 
                <span
                  style="white-space: nowrap;"
                >
                  <span />
                  <span
                    class="mock-TeX"
                  >
                    A=8\\ \\text{in}
                  </span>
                  <span />
                </span>
                , approximately how tall should he make triangle height 
                <span
                  style="white-space: nowrap;"
                >
                  <span />
                  <span
                    class="mock-TeX"
                  >
                    C
                  </span>
                  <span />
                </span>
                ?
              </div>
              <div
                class="sat-grafting-area"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </span>
</div>
`;
